lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2557,11 +2557,14 @@ bool AdjointGen::run() {
 //===----------------------------------------------------------------------===//
 
 namespace {
+class AdjointEmitter;
 
 /// A symbolic adjoint value that is capable of representing zero value 0 and
 /// 1, in addition to a materialized SILValue. This is expected to be passed
 /// around by value in most cases, as it's two words long.
 class AdjointValue {
+  friend class AdjointEmitter;
```

Comment:
Also, make constructors be `protected`.

Reply:
`AdjointValue`'s constructor is currently `private`, so I left it as such (since it's stricter than `protected`).

Reply:
They are obviously still public in the source code.

Reply:
Oops, done in https://github.com/apple/swift/pull/22080/commits/c17276c2b6f8c639f81ea6a218f322245afa85fc.